simple/multi_mr.c

```diff
@@ -0,0 +1,496 @@
+
+			printf("sent successfully\n");
+		}
+	}
+	printf("GOOD, multi mr key test complete\n");
+
+	return ret;
+}
+
+static int init_fabric(void)
+{
+	int ret = 0;
+
+	ret = ft_getinfo(hints, &fi);
+	if (ret)
+		return ret;
+
+	ret = ft_open_fabric_res();
+	if (ret)
+		return ret;
+
+	/* Prevent memory registration by ft_alloc_active_res() -> ft_alloc_msgs() */
+	ft_skip_mr = 1;
+
+	ret = ft_alloc_active_res(fi);
+	if (ret)
+		return ret;
+
+	ret = alloc_multi_mr_res();
+	if (ret)
+		return ret;
+
+	ret = ft_init_ep();
+	if (ret)
+		return ret;
+
+	ret = ft_init_av();
+	if (ret)
+		return ret;
+
+	return ret;
+}
+
+static int server_connect(void)
+{
+	struct fi_eq_cm_entry entry;
+	uint32_t event;
+	ssize_t rd;
+	int ret;
+
+	rd = fi_eq_sread(eq, &event, &entry, sizeof(entry), -1, 0);
+	if (rd != sizeof(entry)) {
+		FT_PROCESS_EQ_ERR(rd, eq, "fi_eq_sread", "listen");
+		ret = (int) rd;
+		return ret;
+	}
+
+	fi = entry.info;
+	if (event != FI_CONNREQ) {
+		fprintf(stderr, "Unexpected CM event %d\n", event);
+		ret = -FI_EOTHER;
+		goto err;
+	}
+
+	ret = fi_domain(fabric, fi, &domain, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_domain", ret);
+		goto err;
+	}
+
+	/* Prevent memory registration by ft_alloc_active_res() -> ft_alloc_msgs() */
+	ft_skip_mr = 1;
+
+	ret = ft_alloc_active_res(fi);
+	if (ret)
+		return ret;
+
+	alloc_multi_mr_res();
+
+	ret = ft_init_ep();
+	if (ret)
+		return ret;
+
+	ret = fi_accept(ep, NULL, 0);
+	if (ret) {
+		FT_PRINTERR("fi_accept", ret);
+		goto err;
+	}
+
+	rd = fi_eq_sread(eq, &event, &entry, sizeof(entry), -1, 0);
+	if (rd != sizeof(entry)) {
+		FT_PROCESS_EQ_ERR(rd, eq, "fi_eq_sread", "accept");
+		ret = (int) rd;
+		goto err;
+	}
+
+	if (event != FI_CONNECTED || entry.fid != &ep->fid) {
+		fprintf(stderr, "Unexpected CM event %d fid %p (ep %p)\n",
+			event, entry.fid, ep);
+		ret = -FI_EOTHER;
+		goto err;
+	}
+
+	return 0;
+
+err:
+	fi_reject(pep, fi->handle, NULL, 0);
+	return ret;
+}
+
+static int client_connect(void)
+{
+	struct fi_eq_cm_entry entry;
+	uint32_t event;
+	ssize_t rd;
+	int ret;
+
+	ret = ft_getinfo(hints, &fi);
+	if (ret)
+		return ret;
+
+	ret = ft_open_fabric_res();
+	if (ret)
+		return ret;
+
+	/* Prevent memory registration by ft_alloc_active_res() -> ft_alloc_msgs() */
+	ft_skip_mr = 1;
+
+	ret = ft_alloc_active_res(fi);
+	if (ret)
+		return ret;
+
+	ret = alloc_multi_mr_res();
+	if (ret)
+		return ret;
+
+	ret = ft_init_ep();
+	if (ret)
+		return ret;
+
+	ret = fi_connect(ep, fi->dest_addr, NULL, 0);
+	if (ret) {
+		FT_PRINTERR("fi_connect", ret);
+		return ret;
+	}
+
+	rd = fi_eq_sread(eq, &event, &entry, sizeof(entry), -1, 0);
+	if (rd != sizeof(entry)) {
+		FT_PROCESS_EQ_ERR(rd, eq, "fi_eq_sread", "connect");
+		ret = (int) rd;
+		return ret;
+	}
+
+	if (event != FI_CONNECTED || entry.fid != &ep->fid) {
+		fprintf(stderr, "Unexpected CM event %d fid %p (ep %p)\n",
+			event, entry.fid, ep);
+		ret = -FI_EOTHER;
+		return ret;
+	}
+
+	return 0;
+}
```

Comment:
The above server_connect, client_connect, and init_fabric calls are almost identical to shared calls.  It looks like the main differences is the call to alloc_multi_mr_res().  There's an existing boolean variable, ft_skip_mr, which is used to skip the standard memory registration path in the common code.

Can you see if we can replace ft_skip_mr with a function callback instead?  If the callback is not set, continue down the standard registration path.  Otherwise have the common code invoke the callback, then continue.  For this test, the callback would be alloc_multi_mr_res().

I'm not sure if this approach will work without investigating the changes more, but it should give us more flexibility in how each test allocates and registers its buffers.